use-cases/Separate-interface-from-implementation.md

```diff
@@ -0,0 +1,29 @@
+# Separate interface from implementation
+
+## General information
+
+The interface and implementation of package definitions should be separated to ensure better maintainability, flexibility, and compatibility between different tools and approaches. A well-defined interface would allow package maintainers to focus on the implementation details while providing stability and consistency for end-users. It would also enable out-of-tree packaging libraries to interoperate with Nixpkgs more effectively, reducing the need for duplicated efforts and custom workarounds.
+
+## Concrete examples
+
+Scenario 1: Out of tree implementations (eg. lang2nix)
+Alice maintains a packaging library that simplifies python packaging. Because nixpkgs offers a well defined packaging interface, she is able to design her library so that it provides good stability without suffering from sudden breakages introduced by implementation changes in nixpkgs.
+
+Scenario 2: Change implementation details
+Bob is a maintainer of the python framework in nixpkgs. He makes an improvement on the python packaging logic. He can be confident about not breaking downstream packages because he only changed the implementation, not the interface.
+
+Scenario 3: Mixing different implementations
+Carl wants to build a python project with dependencies. Some of the dependencies are defined by a lock file and are not available from nixpkgs. Carl would like to build some of the packages using poetry2nix while mixing in other existing package from nixpkgs seamlessly.
+
+## Current problems
+
+**Non-composable implementations**: Different implementations like the one from nixpkgs and out of tree packaging libraries (eg. lang2nix) are often incompatible. They often cannot be easily combined and there is a lock-in to a certain tool. For example, if poetry2nix is used for python, the user cannot simply inherit all the well-maintained build logic from nixpkgs. The logic from nixpkgs has to be duplicated manually, by individual users or the tools community.
```

Comment:
It's not entirely clear what this has to do with interface/implementation separation, since having the latter would not necessarily be composable.

> For example, if poetry2nix is used for python, the user cannot simply inherit all the well-maintained build logic from nixpkgs.

Who is the "user" in this case? A user of poetry2nix, or the implementer of poetry2nix? And why can't they re-use the build logic from Nixpkgs? It makes sense in the case of the implementer, since they wouldn't want to depend on a fragile implementation.

Reply:
I have rewritten and renamed the section in afe4ba1c29a25d1e2908ebdeb22812a242f5820c.
I also added the explanation:
`While having well defined interfaces isn't a guarantee for compatibility, it can make it simpler to achieve it.`

@edolstra does it make more sense now? If not, I could just remove this problem. We have enough other ones anyways.